src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,885 @@
+	 * if not reset their state back to scheduled.
+	 */
+	{
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		ResetRunningBackgroundTasks();
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+
+	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
+	bool hasJobs = true;
+	while (hasJobs)
+	{
+		MemoryContextReset(perTaskContext);
+
+		CHECK_FOR_INTERRUPTS();
+
+		InvalidateMetadataSystemCache();
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * We need to load the task into the perTaskContext as we will switch contexts
+		 * later due to the committing and starting of new transactions
+		 */
+		MemoryContext oldContext = MemoryContextSwitchTo(perTaskContext);
+		BackgroundTask *task = GetRunnableBackgroundTask();
+		MemoryContextSwitchTo(oldContext);
+
+		if (!task)
+		{
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+
+			hasJobs = false;
+			break;
+		}
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		/* TODO find the actual database and username */
+		dsm_segment *seg = NULL;
+		BackgroundWorkerHandle *handle =
+			StartCitusBackgroundJobExecuter("postgres", "nilsdijk", task->command,
+											&seg);
+
+		if (handle == NULL)
+		{
+			/* TODO something better here */
+			ereport(ERROR, (errmsg("unable to start background worker")));
+		}
+
+		pid_t pid = 0;
+		GetBackgroundWorkerPid(handle, &pid);
+
+		ereport(LOG, (errmsg("found task with jobid: %ld", task->taskid)));
+
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		task->status = BACKGROUND_TASK_STATUS_RUNNING;
+		BackgroundTaskUpdatePid(task, &pid);
+
+		/* Update task status to indicate it is running */
+		UpdateBackgroundTask(task);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		bool hadError = false;
+
+		/*
+		 * We reset the old message (if present). This will only retain the last message
+		 * in the catalog. Otherwise it would concatenate all retries.
+		 */
+		if (task->message)
+		{
+			pfree(task->message);
+		}
+		task->message = NULL;
+
+		/* TODO keep polling the task */
+		while (GetBackgroundWorkerPid(handle, &pid) != BGWH_STOPPED)
+		{
+			int latchFlags = WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH;
+			int rc = WaitLatch(MyLatch, latchFlags, (long) 1000, PG_WAIT_EXTENSION);
+
+			/* emergency bailout if postmaster has died */
+			if (rc & WL_POSTMASTER_DEATH)
+			{
+				proc_exit(1);
+			}
+
+			if (rc & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+				CHECK_FOR_INTERRUPTS();
+			}
+
+			shm_toc *toc = shm_toc_attach(CITUS_BACKGROUND_TASK_MAGIC,
+										  dsm_segment_address(seg));
+			shm_mq *mq = shm_toc_lookup(toc, CITUS_BACKGROUND_TASK_KEY_QUEUE, false);
+			shm_mq_handle *responseq = shm_mq_attach(mq, seg, NULL);
+
+			ConsumeTaskWorkerOutput(responseq, task, &hadError);
```

Comment:
this function allocates memory which will leak into perTaskContext

Reply:
The `perTaskContext` should be reset between two task runs. I think the leak is ok, or do you think explicitly free-ing the leaking memory?